Automattic/wp-calypso
# chore: migration to RQ v5 (major)

Related to #84413

⚠️ This PR should merge into #84413 and **not** into `trunk`.

See the official migration guide for React Query v5: https://tanstack.com/query/latest/docs/react/guides/migrating-to-v5

## Upgrade path: breaking changes

- [x] Always use object format for `useQuery` and friends ([link](https://tanstack.com/query/latest/docs/react/guides/migrating-to-v5#supports-a-single-signature-one-object))
  - https://github.com/Automattic/wp-calypso/pull/84378
  - https://github.com/Automattic/wp-calypso/pull/84380
  - https://github.com/Automattic/wp-calypso/pull/84416
  - #84471
  - https://github.com/Automattic/wp-calypso/pull/86114
- [x] Dont use `onSuccess`, `onError` and `onSettled` with `useQuery` ([link](https://tanstack.com/query/latest/docs/react/guides/migrating-to-v5#callbacks-on-usequery-and-queryobserver-have-been-removed))
  - #84415
  - #84464
  -  #86118
  - #86149
  - see https://github.com/Automattic/wp-calypso/pull/84338#issuecomment-1824525840
- [x] ~~`refetchInterval`: use `query.state.data` to retrieve `data` ([link](https://tanstack.com/query/latest/docs/react/guides/migrating-to-v5#the-refetchinterval-callback-function-only-gets-query-passed))~~ no usage found where we pass a function
- [x] Remove any usages of the deprecated logger ([link](https://tanstack.com/query/latest/docs/react/guides/migrating-to-v5#callbacks-on-usequery-and-queryobserver-have-been-removed))
- [x] Rename `cacheTime` to `gcTime` ([link](https://tanstack.com/query/latest/docs/react/guides/migrating-to-v5#rename-cachetime-to-gctime))
- [x] Replace `keepPreviousData` with `placeholderData: keepPreviousData` ([link](https://tanstack.com/query/latest/docs/react/guides/migrating-to-v5#removed-keeppreviousdata-in-favor-of-placeholderdata-identity-function))
- [x] Infinite queries need `initialPageParam` ([link](https://tanstack.com/query/latest/docs/react/guides/migrating-to-v5#infinite-queries-now-need-a-initialpageparam))
- [x] Fix type errors related to `queryKey` not being excluded from `UserQueryOptions` on custom query hooks
  - #84472
  - #84600
- [x] Fix type errors related to `TError` generic changing to `Error` instead of `unknown`
  -  #84474
  - #84599
- [ ] Rename `isLoading` to `isPending` ([link](https://tanstack.com/query/latest/docs/react/guides/migrating-to-v5#status-loading-has-been-changed-to-status-pending-and-isloading-has-been-changed-to-ispending-and-isinitialloading-has-now-been-renamed-to-isloading))
	- [x] `isLoading` usages on mutations
	- [ ] `isLoading` usages on queries
	- note: `isInitialLoading` is deprecated and being renamed to `isLoading` which is the same as `isPending && isFetching` – we'll keep this for now and follow up
  - #84465

## Testing Instructions

- Smoke test Calypso and thoroughly test important flows like login, signup, etc – everything should behave exactly the same as on prod
- Verify there are no type errors and no eslint failures
- Verify all tests are passing and failing tests are fixed in a reasonable way

## Pre-merge Checklist
